### PR TITLE
detect incorrect encoding names for mb_functions

### PR DESCRIPTION
**Pull request by Danil42Russia**
#### Table of behavioral differences in PHP / KPHP

| Name         | PHP | KPHP | Comment                                                                          |
|:-------------|:---:|:----:|:---------------------------------------------------------------------------------|
| cp1251       |  ✅  |  ✅   | It's okay, it accepts it as **windows-1251**                                     |
| utf-8        |  ✅  |  ✅   | It's okay, it accepts it as **UTF-8**                                            |
| utf8         |  ✅  |  ❌   | In KPHP error `encoding "utf8" doesn't supported"`, PHP accepts as **UTF-8**     |
| windows1251  |  ❌  |  ✅   | In PHP error `Invalid encoding "windows1251"`, KPHP accepts as **windows-1251**  |
| lol-1251-kek |  ❌  |  ✅   | In PHP error `Invalid encoding "lol-1251-kek"`, KPHP accepts as **windows-1251** |
| lol-8-kek    |  ❌  |  ✅   | In PHP error `Invalid encoding "lol-8-kek"`, KPHP accepts as **UTF-8**           |

#### Alias list in PHP

1. [windows-1251](https://github.com/php/php-src/blob/cebb8009c6bebd8af6c5cbdc803d632ae1252eeb/ext/mbstring/libmbfl/filters/mbfilter_singlebyte.c#L465)
2. [UTF-8](https://github.com/php/php-src/blob/cebb8009c6bebd8af6c5cbdc803d632ae1252eeb/ext/mbstring/libmbfl/filters/mbfilter_utf8.c#L55)

The changes will take place in several stages:
1. Search for incorrect encoding names (**we are here now**)
2. Replacement of encoding evaluation check

### CODE REVIEW

File: runtime/mbstring.cpp [hunk 1]
```diff
@@ -7,7 +7,31 @@
 #include "common/unicode/unicode-utils.h"
 #include "common/unicode/utf8-utils.h"
 
+bool is_detect_incorrect_encoding_names_warning = false;
+
+void f$set_detect_incorrect_encoding_names_warning(bool show) {
+  is_detect_incorrect_encoding_names_warning = show;
+}
+
+static void detect_incorrect_encoding_names(const string &encoding) {
```
Comment (unserialize) [inline runtime/mbstring.cpp:16]: This function should be `static int mb_detect_encoding_new()` and return a number. It should not fire any error

File: runtime/mbstring.cpp [hunk 2]
```diff
@@ -7,7 +7,31 @@
...
+  }
+
+  php_warning("Found unsupported \"%s\" encoding", encoding_name);
+}
+
 static int mb_detect_encoding(const string &encoding) {
+  if (is_detect_incorrect_encoding_names_warning) {
+    detect_incorrect_encoding_names(encoding);
+  }
+
```
Comment (unserialize) [inline runtime/mbstring.cpp:34]: These lines should not exist

File: runtime/mbstring.cpp [hunk 3]
```diff
@@ -7,7 +7,31 @@
...
   if (strstr(encoding.c_str(), "1251")) {
     return 1251;
```
Comment (unserialize) [inline runtime/mbstring.cpp:36]: Before returning, it should check equality:
```
if (1251 != mb_detect_encoding_new(encoding) && is_detect_incorrect_encoding_names_warning) {
  php_warning(...);
}
```

same before other returns inside this function.

File: runtime/mbstring.cpp [hunk 4]
```diff
@@ -7,7 +7,31 @@
 #include "common/unicode/unicode-utils.h"
 #include "common/unicode/utf8-utils.h"
 
+bool is_detect_incorrect_encoding_names_warning = false;
```
Comment (unserialize) [inline runtime/mbstring.cpp:10]: I don't see where this variable is reset when the PHP script finishes

File: runtime/mbstring.cpp [hunk 5]
```diff
@@ -7,7 +7,31 @@
...
+    return 8;
+  }
+
+  return -1;
+}
+
 static int mb_detect_encoding(const string &encoding) {
+  if (-1 == mb_detect_encoding_new(encoding) && is_detect_incorrect_encoding_names_warning) {
+    php_warning("Found unsupported \"%s\" encoding", encoding.c_str());
+  }
```
Comment (unserialize) [inline runtime/mbstring.cpp:33]: This is wrong, see my older comments

**Review by unserialize — APPROVED** [review]

**Review by DrDet — APPROVED** [review]

**Pull request merged**